File: ldes-consumer/kgap/spawn_instances.py
#!/usr/bin/env python3
"""
LDES Consumer Spawner
Reads a YAML configuration file and spawns ldes2sparql Docker container instances.
"""
import os
import sys
import yaml
import subprocess
import signal
import time
from pathlib import Path
from contextlib import contextmanager
from typing import Generator, Any
from logger import setup_logger, Logger

# === Global config / data and logger setup ===
# Set up logger
loglevel: str = os.getenv("LOG_LEVEL", "INFO")
pfx: str = os.getenv("LDES_CONSUMER_PREFIX", "ldes-consumer")
log: Logger = setup_logger(pfx, loglevel)

# Get configuration options
# note that js based ldes2sparql expects lowercase log levels
ldes_loglevel = os.getenv("LDES_LOG_LEVEL", loglevel).lower()
image_name: str = os.getenv(
    "LDES2SPARQL_IMAGE", "ghcr.io/maregraph-eu/ldes2sparql:latest"  # default image
)
project_name: str = os.getenv("COMPOSE_PROJECT_NAME", "kgap")  # default kgap
gdb_repo: str = os.getenv("GDB_REPO", "kgap")  # default kgap
default_sparql_endpoint: str = os.getenv(
    "DEFAULT_SPARQL_ENDPOINT", f"http://graphdb:7200/repositories/{gdb_repo}/statements"
)  # default endpoint for updates
network_name: str = os.getenv(
    "DOCKER_NETWORK", f"{project_name}_default"
)  # default network name derived from compose project name
remove_containers: bool = (
    os.getenv("LDES_REMOVE_CONTAINERS", "1") == "1"
)  # default to true
monitor_interval: int = int(
    os.getenv("LDES_MONITOR_INTERVAL", "300")
)  # in seconds, default 5 minutes

host_pwd: str = os.getenv(
    "HOST_PWD", "/tmp"
)  # passed working dir from host env - fallback to /tmp

# path conversion setup between guest and host
guest_data_root: Path = Path("/data")  # main mounted /data inside docker guest
host_data_path: Path = Path(host_pwd) / "data"  # assumed corresponding host data path


def guest2host_data_path(path: Path) -> Path:
    """Convert a guest data path to the corresponding host data path."""
    return host_data_path / path.relative_to(guest_data_root)


# paths in use
ldes_consumer_root: Path = (
    guest_data_root / "ldes-consumer"
)  # main ldes-consumer data path in guest
logs_path: Path = ldes_consumer_root / "logs"  # subfolder for logs in guest
state_path: Path = ldes_consumer_root / "state"  # subfolder for state in guest
host_state_path: Path = guest2host_data_path(
    state_path
)  # corresponding host path for state (for volume mount)

# Global list to track feeds in use
feeds: dict[str, dict] = None


# === Docker container management functions ===
def docker_container_name(feedname: str) -> str:
    """Generate the Docker container name for a given feed."""
    return f"{pfx}-{feedname}"


@contextmanager
def check_docker_container_running(
    feedname: str, feed: dict
) -> Generator[bool, None, None]:
    """Check if a Docker container for a given feed is running."""
    container_name = docker_container_name(feedname)
    cmd = ["docker", "inspect", "-f", "{{.State.Running}}", container_name]
    try:
        result = subprocess.run(cmd, capture_output=True, text=True, timeout=5)
        is_running = result.returncode == 0 and result.stdout.strip() == "true"
        yield is_running
    except subprocess.TimeoutExpired:
        log.error(f"Timeout checking container status for '{container_name}'")
        yield False


def _check_docker_container_exists(feedname: str, feed: dict) -> bool:
    """Check if a Docker container for a given feed exists."""
    container_name = docker_container_name(feedname)
    cmd = ["docker", "inspect", container_name]
    try:
        result = subprocess.run(cmd, capture_output=True, text=True, timeout=5)
        return result.returncode == 0
    except subprocess.TimeoutExpired:
        log.error(f"Timeout checking container existence for '{container_name}'")
        return False


def docker_container_remove(feedname: str, feed: dict) -> None:
    """Remove a Docker container for a given feed if it exists.
    But only if it exists and the wrapper is configured to do so."""
    if not _check_docker_container_exists(feedname, feed):
        log.info(f"No existing container to remove for feed '{feedname}'")
        return
    # else
    if not remove_containers:
        log.info(
            f"Skipping removal of existing container for feed '{feedname}' as per configuration."
        )
        log.info(
            "This may lead to (1) not launching now if it is already running, or (2) reusing an old stopped container."
        )
        return
    # else - proceed to remove
    log.info(f"Removing existing container for feed '{feedname}'")
    container_name = docker_container_name(feedname)
    cmd: list[str] = ["docker", "container", "remove", "-f", container_name]
    try:
        subprocess.run(cmd, timeout=10)
        log.info(f"Removed container '{container_name}' successfully.")
    except Exception as e:
        log.error(f"Error removing container '{container_name}'")
        log.exception(e, exc_info=True)


def docker_container_start(
    feedname: str, feed: dict, image_name: str, project_name: str, network_name: str
) -> None:
    """Start a Docker container for a given feed."""
    # 1| get relevant parts from feed config
    feed_url = feed.get("url")
    sparql_endpoint = feed.get("sparql_endpoint", default_sparql_endpoint)
    target_graph = feed.get("target_graph", f"urn:kgap:{pfx}:{feedname}")
    container_name = docker_container_name(feedname)
    # Note: The config uses 'polling_interval' (seconds) for user-friendliness,
    # but ldes2sparql expects 'POLLING_FREQUENCY' (milliseconds)
    polling_frequency = feed.get("polling_interval", 60) * 1000

    # 2| Compose build the docker run command
    cmd: list[str] = [
        "docker",
        "run",
        "--name",
        container_name,
        "--network",
        network_name,
        # Add labels to link container to the docker-compose project
        "--label",
        f"com.docker.compose.project={project_name}",
        "--label",
        "com.docker.compose.service=ldes-consumer",
        "-v",
        f"{host_data_path}:/data",
        "-v",
        f"{host_state_path}/{feedname}:/state",
    ]

    # Add environment variables
    cmd.extend(["-e", f"LDES={feed_url}"])
    cmd.extend(["-e", f"SPARQL_ENDPOINT={sparql_endpoint}"])
    cmd.extend(["-e", f"TARGET_GRAPH={target_graph}"])
    cmd.extend(["-e", f"POLLING_FREQUENCY={polling_frequency}"])
    cmd.extend(["-e", "FAILURE_IS_FATAL=false"])
    cmd.extend(["-e", "FOLLOW=true"])

    # Add any additional environment variables from the feed config
    added_envs: set[str] = set()
    extra_env = feed.get("environment", {})
    if isinstance(extra_env, dict):
        for key, value in extra_env.items():
            if key in {
                "LDES",
                "SPARQL_ENDPOINT",
                "TARGET_GRAPH",
                "FAILURE_IS_FATAL",
                "FOLLOW",
                "POLLING_FREQUENCY",
            }:
                log.warning(
                    f"Environment variable '{key}' for feed '{feedname}' is reserved and cannot be overridden; ignoring..."
                )
                continue
            cmd.extend(["-e", f"{key}={value}"])
            added_envs.add(key)
    else:
        log.warning(
            f"'environment' for feed '{feedname}' is not a mapping; ignoring..."
        )
    # Ensure OPERATION_MODE is set, default to 'sync' if not provided
    if "OPERATION_MODE" not in added_envs:
        cmd.extend(["-e", "OPERATION_MODE=Sync"])
    # Ensure MAXIMUM_BATCH_SIZE is set, default to 500 if not provided
    if "MEMBER_BATCH_SIZE" not in added_envs:
        cmd.extend(["-e", "MEMBER_BATCH_SIZE=500"])
    # Ensure SHAPE is set, even if empty
    if "SHAPE" not in added_envs:
        cmd.extend(["-e", "SHAPE="])
    # Ensure MATERIALIZE is set, even if empty
    if "MATERIALIZE" not in added_envs:
        cmd.extend(["-e", "MATERIALIZE=true"])
    # Ensure LOG_LEVEL is set correctly
    if "LOG_LEVEL" not in added_envs:
        cmd.extend(["-e", f"LOG_LEVEL={ldes_loglevel}"])

    # Add the image name
    cmd.append(image_name)

    # 3| Prepare state directory
    state_path_for_feed = state_path / feedname
    state_path_for_feed.mkdir(parents=True, exist_ok=True)
    json_state_file = state_path_for_feed / "ldes-client_state.json"
    if json_state_file.exists():
        log.info(
            f"Note that existing state file exists for feed '{feedname}' it will be reused."
        )

    log.info(f"Starting LDES consumer for feed: {feedname}")
    log.info(f"  URL: {feed_url}")
    log.info(f"  SPARQL Endpoint: {sparql_endpoint}")
    log.info(f"  Target Graph: {target_graph}")
    log.info(f"  Polling Frequency (ms): {polling_frequency}")
    log.info(f"  State path: {host_state_path}/{feedname}")
    log.debug(f"Starting LDES feed instance with command:\n{' '.join(cmd)}")

    # 4| Run the docker container
    try:
        proc: subprocess.Popen = subprocess.Popen(
            cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True
        )
        # Give the container a moment to start
        time.sleep(2)

        # Check if the container is actually running
        with check_docker_container_running(feedname, feed) as is_running:
            if is_running:
                active_feed(feedname, feed, proc)
                return
            # else
            log.error(f"Container '{feedname}' failed to start properly.")
            docker_container_capture_logs(feedname, feed)
            docker_container_remove(feedname, feed)
            fail_feed(feedname, feed, "Container failed to start.")
            return

    except Exception as e:
        log.error(f"Failed to spawn container for feed '{feedname}'")
        log.exception(e, exc_info=True)
        fail_feed(feedname, feed, "Exception during container start.")
    return


def docker_container_stop(feedname: str, feed: dict) -> bool:
    """Stop a Docker container for a given feed."""
    container_name = docker_container_name(feedname)
    cmd: list[str] = ["docker", "stop", container_name]
    try:
        subprocess.run(cmd, timeout=60)
        log.info(f"Terminated process for feed '{feedname}'")
    except subprocess.TimeoutExpired:
        log.error(f"Timeout while trying to stop docker for feed '{feedname}'")
    except Exception as e:
        log.error(f"Error terminating process for feed '{feedname}'")
        log.exception(e, exc_info=True)


def docker_container_capture_logs(feedname: str, feed: dict) -> None:
    """Capture logs from a Docker container for a given feed."""
    container_name = docker_container_name(feedname)
    # Try to get logs
    cmd: list[str] = ["docker", "logs", container_name]
    # allocate filenames to capture stdout and stderr
    ts = time.strftime("%Y%m%d-%H%M%S")
    stdout_log = logs_path / f"{feedname}_{ts}_stdout.log"
    stderr_log = logs_path / f"{feedname}_{ts}_stderr.log"

    with open(stdout_log, "w") as stdout_file, open(stderr_log, "w") as stderr_file:
        capture_logs = subprocess.run(
            cmd, stdout=stdout_file, stderr=stderr_file, timeout=5
        )
        if capture_logs.returncode == 0:
            log.error(
                f"Container logs can be found in {logs_path}/{feedname}_{ts}*.log"
            )


# === Signal handling for graceful shutdown ===
def signal_handler(signum, frame):
    """Handle shutdown signals gracefully to stop all spawned docker images for individual feeds."""
    global feeds
    log.info(f"Received signal {signum}, checking {len(feeds)} LDES consumers...")
    active_feeds: dict[str, dict] = get_active_feeds()
    log.info(
        f"Processing signal {signum}, shutting down {len(active_feeds)} active LDES consumers..."
    )
    for feedname, feed in active_feeds.items():
        with check_docker_container_running(feedname, feed) as is_running:
            if not is_running:
                log.info(
                    f"active LDES consumer for feed '{feedname}' not running, so not stopping"
                )
            else:
                log.info(f"Stopping active LDES consumer for feed '{feedname}'...")
                docker_container_stop(feedname, feed)
        docker_container_remove(feedname, feed)
    log.info(
        f"Done handling signal {signum}, all feed instances should have stopped..."
    )
    sys.stdout.flush()
    sys.exit(0)


# === Configuration loading and feed spawning logic ===
def load_config(feed_config_path: Path) -> dict[str, Any]:
    """Load and parse the YAML configuration file."""
    try:
        with open(feed_config_path, "r") as feed_config_file:
            feed_config = yaml.safe_load(feed_config_file)
        return feed_config
    except Exception as e:
        log.error(f"Failed to load config file {feed_config_path}")
        log.exception(e, exc_info=True)
        log.info("Exiting due to configuration load failure.")
        sys.exit(1)


def fail_feed(feedname: str, feed: dict, reason: str) -> None:
    """Helper to log feed failure reasons."""
    log.error(f"Feed {feedname} setup failed: {reason}")
    feed["active"] = False
    feed["failure_reason"] = reason
    return


def active_feed(feedname: str, feed: dict, proc: subprocess.Popen) -> None:
    """Helper to log feed success."""
    log.info(f"Feed {feedname} docker instance started successfully.")
    feed["active"] = True
    feed["process"] = proc
    feed.pop("failure_reason", None)  # remove any previous failure reason
    return


def get_active_feeds() -> dict[str, dict]:
    """Get a dictionary of currently active feeds."""
    global feeds
    return {fname: f for fname, f in feeds.items() if f.get("active", True)}


def spawn_feed_instance(
    feedname: str,
    feed: dict[str, Any],
    image_name: str,
    project_name: str,
    network_name: str,
) -> None:
    """
    Spawn a single ldes2sparql Docker container instance.

    Args:
        feedname: unique name of the feed
        feed: dictionary containing feed configuration
        image_name: ref to docker image to use for ldes2sparql
        project_name: docker-compose project name for container labeling
        network_name: docker network to attach to in order to reach triple store (graphdb)

    Returns:
        Nothing. Logs errors if spawning fails. Adds flag 'active' flag to the feed dict on success.
    """

    if "url" not in feed:
        fail_feed(feedname, feed, "Missing URL in feed configuration")
        return

    with check_docker_container_running(feedname, feed) as is_running:
        if is_running:
            fail_feed(
                feedname,
                feed,
                "Container is already running. Stop it or give a different name.",
            )
            return

    docker_container_remove(feedname, feed)
    docker_container_start(feedname, feed, image_name, project_name, network_name)
    return


# === Main execution logic ===
def main():
    """Main function to spawn all ldes2sparql instances."""
    log.info(f"Spawner started with {len(sys.argv)} cli arguments")
    log.info(f"Logger initialized with level {loglevel}")

    if len(sys.argv) < 2:
        log.error("Usage: spawn_instances.py <config_file>")
        sys.exit(1)

    config_file: str = sys.argv[1]

    # Set up signal handlers for graceful shutdown
    signal.signal(signal.SIGTERM, signal_handler)
    signal.signal(signal.SIGINT, signal_handler)

    # Load configuration
    config = load_config(config_file)

    # Populate feeds list
    global feeds
    feeds = config.get("feeds", [])
    if not feeds:
        log.error("No feeds defined in configuration file")
        sys.exit(1)

    log.info(f"Found {len(feeds)} LDES feed(s) to process")

    for feedname, feed in feeds.items():
        spawn_feed_instance(
            feedname,
            feed,
            image_name,
            project_name,
            network_name,
        )

    active_feeds: dict[str, dict] = get_active_feeds()

    if not active_feeds:
        log.error(f"No LDES consumers (out of {len(feeds)}) were started successfully")
        sys.exit(1)

    log.info(
        f"Successfully started {len(active_feeds)} of {len(feeds)} LDES consumer(s)"
    )
    log.info("Starting to monitor started processes... (Press Ctrl+C to stop)")

    # Monitor processes and restart if they have ended
    while True:
        time.sleep(monitor_interval)
        active_feeds: dict[str, dict] = get_active_feeds()

        log.info(f"Monitoring of {len(active_feeds)} LDES consumer(s)...")
        for feedname, feed in active_feeds.items():
            log.info(f"Checking LDES consumer for feed '{feedname}'...")
            with check_docker_container_running(feedname, feed) as is_running:
                if is_running:
                    log.info(f"LDES consumer for feed '{feedname}' is still running")
                    continue  # still running
                # else - container has stopped - attempt restart
                log.warning(
                    f"LDES consumer for feed '{feedname}' has stopped - capturing logs and attempting restart"
                )
                docker_container_capture_logs(feedname, feed)
                docker_container_start(
                    feedname,
                    feed,
                    image_name,
                    project_name,
                    network_name,
                )


if __name__ == "__main__":
    main()

File: ldes-consumer/kgap/spawn_instances.js
#!/usr/bin/env node
/**
 * LDES Consumer Spawner
 * Manages spawning and monitoring of ldes2sparql container instances
 */

const fs = require('fs');
const path = require('path');
const { exec, spawn } = require('child_process');
const { promisify } = require('util');
const execAsync = promisify(exec);

// Configuration
const LDES_CONFIG_PATH = process.env.LDES_CONFIG_PATH || '';
const DOCKER_NETWORK = process.env.DOCKER_NETWORK || 'kgap_default';
const COMPOSE_PROJECT_NAME = process.env.COMPOSE_PROJECT_NAME || 'kgap';
const LOG_LEVEL = process.env.LOG_LEVEL || 'INFO';
const LOCK_FILE = path.join(LDES_CONFIG_PATH, '.spawn.lock');

// Track active containers
const activeContainers = new Map();

// Logging functions
function log(level, ...args) {
    const timestamp = new Date().toISOString().replace('T', ' ').substring(0, 19);
    const levels = { DEBUG: 0, INFO: 1, WARNING: 2, ERROR: 3 };
    const currentLevel = levels[LOG_LEVEL] || 1;
    
    if (levels[level] >= currentLevel) {
        console.log(`${timestamp} - ldes-spawner - ${level} -`, ...args);
    }
}

// Get own container info from hostname
async function getOwnContainerInfo() {
    try {
        const hostname = require('os').hostname();
        const { stdout } = await execAsync(`docker inspect ${hostname}`);
        const info = JSON.parse(stdout)[0];
        return {
            id: info.Id,
            name: info.Name.replace('/', ''),
            image: info.Config.Image
        };
    } catch (error) {
        log('ERROR', 'Failed to get own container info:', error.message);
        throw error;
    }
}

// Acquire lock to prevent competing spawning instances
async function acquireLock() {
    const maxRetries = 10;
    const retryDelay = 1000;
    
    for (let i = 0; i < maxRetries; i++) {
        try {
            // Try to create lock file exclusively
            fs.writeFileSync(LOCK_FILE, process.pid.toString(), { flag: 'wx' });
            log('INFO', 'Lock acquired');
            return true;
        } catch (error) {
            if (error.code === 'EEXIST') {
                // Lock file exists, check if process is still running
                try {
                    const pid = parseInt(fs.readFileSync(LOCK_FILE, 'utf8'));
                    try {
                        process.kill(pid, 0); // Check if process exists
                        log('DEBUG', `Lock held by process ${pid}, waiting...`);
                        await new Promise(resolve => setTimeout(resolve, retryDelay));
                    } catch {
                        // Process doesn't exist, remove stale lock
                        log('WARNING', 'Removing stale lock file');
                        fs.unlinkSync(LOCK_FILE);
                    }
                } catch {
                    // Can't read lock file, remove it
                    try {
                        fs.unlinkSync(LOCK_FILE);
                    } catch {}
                }
            } else {
                throw error;
            }
        }
    }
    
    throw new Error('Failed to acquire lock after maximum retries');
}

// Release lock
function releaseLock() {
    try {
        if (fs.existsSync(LOCK_FILE)) {
            fs.unlinkSync(LOCK_FILE);
            log('INFO', 'Lock released');
        }
    } catch (error) {
        log('ERROR', 'Failed to release lock:', error.message);
    }
}

// Parse YAML file (simple key-value extraction)
function parseYamlValue(content, key, defaultValue = '') {
    const regex = new RegExp(`^${key}:\\s*(.*)$`, 'm');
    const match = content.match(regex);
    if (!match) return defaultValue;
    
    let value = match[1].trim();
    // Remove quotes if present
    value = value.replace(/^["']|["']$/g, '');
    return value;
}

// Load and validate YAML config
function loadYamlConfig(filePath) {
    try {
        const content = fs.readFileSync(filePath, 'utf8');
        const url = parseYamlValue(content, 'url');
        const sparqlEndpoint = parseYamlValue(content, 'sparql_endpoint');
        
        if (!url || !sparqlEndpoint) {
            log('ERROR', `YAML file ${filePath} missing required fields 'url' or 'sparql_endpoint'`);
            return null;
        }
        
        return {
            url,
            sparql_endpoint: sparqlEndpoint,
            shape: parseYamlValue(content, 'shape', ''),
            target_graph: parseYamlValue(content, 'target_graph', ''),
            follow: parseYamlValue(content, 'follow', 'true'),
            materialize: parseYamlValue(content, 'materialize', 'false'),
            order: parseYamlValue(content, 'order', 'none'),
            last_version_only: parseYamlValue(content, 'last_version_only', 'false'),
            failure_is_fatal: parseYamlValue(content, 'failure_is_fatal', 'false'),
            polling_interval: parseInt(parseYamlValue(content, 'polling_interval', '5')),
            concurrent_fetches: parseInt(parseYamlValue(content, 'concurrent_fetches', '10')),
            for_virtuoso: parseYamlValue(content, 'for_virtuoso', 'false'),
            query_timeout: parseInt(parseYamlValue(content, 'query_timeout', '1800')),
            before: parseYamlValue(content, 'before', ''),
            after: parseYamlValue(content, 'after', ''),
            access_token: parseYamlValue(content, 'access_token', ''),
            perf_name: parseYamlValue(content, 'perf_name', '')
        };
    } catch (error) {
        log('ERROR', `Failed to load YAML file ${filePath}:`, error.message);
        return null;
    }
}

// Generate container name from filename
function getContainerName(filename) {
    let name = filename.replace(/\.(yaml|yml)$/, '');
    name = name.replace(/[_ ]/g, '-').toLowerCase();
    return `ldes-consumer-${name}`;
}

// Spawn a container from YAML config
async function spawnContainer(yamlPath, ownImage) {
    const filename = path.basename(yamlPath);
    
    if (activeContainers.has(filename)) {
        log('WARNING', `Container already exists for ${filename}`);
        return false;
    }
    
    const config = loadYamlConfig(yamlPath);
    if (!config) return false;
    
    const containerName = getContainerName(filename);
    const pollingFrequency = config.polling_interval * 1000;
    
    log('INFO', `Starting LDES consumer from config file: ${filename}`);
    log('INFO', `  Container name: ${containerName}`);
    log('INFO', `  URL: ${config.url}`);
    log('INFO', `  SPARQL Endpoint: ${config.sparql_endpoint}`);
    
    // Build docker run command
    const dockerArgs = [
        'run', '-d',
        '--name', containerName,
        '--network', DOCKER_NETWORK,
        '--label', `com.docker.compose.project=${COMPOSE_PROJECT_NAME}`,
        '--label', 'com.docker.compose.service=ldes-consumer',
        '--label', `ldes.config.file=${filename}`,
        '-v', `/data/ldes-state-${containerName}:/state`,
        '-e', `LDES=${config.url}`,
        '-e', `SPARQL_ENDPOINT=${config.sparql_endpoint}`,
        '-e', `SHAPE=${config.shape}`,
        '-e', `TARGET_GRAPH=${config.target_graph}`,
        '-e', `FOLLOW=${config.follow}`,
        '-e', `MATERIALIZE=${config.materialize}`,
        '-e', `ORDER=${config.order}`,
        '-e', `LAST_VERSION_ONLY=${config.last_version_only}`,
        '-e', `FAILURE_IS_FATAL=${config.failure_is_fatal}`,
        '-e', `POLLING_FREQUENCY=${pollingFrequency}`,
        '-e', `CONCURRENT_FETCHES=${config.concurrent_fetches}`,
        '-e', `FOR_VIRTUOSO=${config.for_virtuoso}`,
        '-e', `QUERY_TIMEOUT=${config.query_timeout}`
    ];
    
    // Add optional environment variables
    if (config.before) dockerArgs.push('-e', `BEFORE=${config.before}`);
    if (config.after) dockerArgs.push('-e', `AFTER=${config.after}`);
    if (config.access_token) dockerArgs.push('-e', `ACCESS_TOKEN=${config.access_token}`);
    if (config.perf_name) dockerArgs.push('-e', `PERF_NAME=${config.perf_name}`);
    
    // Use own image
    dockerArgs.push(ownImage);
    
    log('DEBUG', 'Docker command:', 'docker', dockerArgs.join(' '));
    
    try {
        await execAsync(`docker ${dockerArgs.join(' ')}`);
        
        // Wait for container to start
        await new Promise(resolve => setTimeout(resolve, 2000));
        
        // Verify container is running
        const { stdout } = await execAsync(`docker inspect -f '{{.State.Running}}' ${containerName}`);
        if (stdout.trim() === 'true') {
            log('INFO', `Successfully started container: ${containerName}`);
            activeContainers.set(filename, containerName);
            return true;
        } else {
            log('ERROR', `Container ${containerName} failed to start properly`);
            const { stdout: logs } = await execAsync(`docker logs ${containerName}`);
            log('ERROR', 'Container logs:', logs.substring(0, 500));
            return false;
        }
    } catch (error) {
        log('ERROR', `Failed to start container ${containerName}:`, error.message);
        return false;
    }
}

// Stop a container
async function stopContainer(filename) {
    const containerName = activeContainers.get(filename);
    
    if (!containerName) {
        log('WARNING', `No active container found for ${filename}`);
        return false;
    }
    
    log('INFO', `Stopping container: ${containerName}`);
    
    try {
        await execAsync(`docker stop ${containerName}`);
        await execAsync(`docker rm ${containerName}`).catch(() => {});
        activeContainers.delete(filename);
        log('INFO', `Successfully stopped and removed container: ${containerName}`);
        return true;
    } catch (error) {
        log('ERROR', `Failed to stop container ${containerName}:`, error.message);
        return false;
    }
}

// Watch folder for changes
async function watchFolder(folder, ownImage) {
    log('INFO', `Scanning directory for YAML files: ${folder}`);
    
    // Initial scan - start containers for existing files
    const files = fs.readdirSync(folder)
        .filter(f => f.endsWith('.yaml') || f.endsWith('.yml'))
        .filter(f => f !== '.spawn.lock');
    
    for (const file of files) {
        await spawnContainer(path.join(folder, file), ownImage);
    }
    
    log('INFO', 'File watcher started. Monitoring for changes...');
    
    // Watch for file changes
    const watcher = fs.watch(folder, async (eventType, filename) => {
        if (!filename || (!filename.endsWith('.yaml') && !filename.endsWith('.yml'))) {
            return;
        }
        
        if (filename === '.spawn.lock') return;
        
        const filePath = path.join(folder, filename);
        
        // Wait a bit for file operations to complete
        await new Promise(resolve => setTimeout(resolve, 1000));
        
        if (eventType === 'rename') {
            // File created or deleted
            if (fs.existsSync(filePath)) {
                log('INFO', `Detected new YAML file: ${filename}`);
                await spawnContainer(filePath, ownImage);
            } else {
                log('INFO', `Detected deleted YAML file: ${filename}`);
                await stopContainer(filename);
            }
        } else if (eventType === 'change') {
            log('INFO', `Detected modified YAML file: ${filename}`);
            // Stop existing container and restart with new config
            if (activeContainers.has(filename)) {
                await stopContainer(filename);
            }
            if (fs.existsSync(filePath)) {
                await spawnContainer(filePath, ownImage);
            }
        }
    });
    
    // Monitor container health
    setInterval(async () => {
        for (const [filename, containerName] of activeContainers.entries()) {
            try {
                const { stdout } = await execAsync(`docker inspect -f '{{.State.Running}}' ${containerName}`);
                if (stdout.trim() !== 'true') {
                    log('WARNING', `Container ${containerName} is not running, restarting...`);
                    activeContainers.delete(filename);
                    
                    const yamlPath = path.join(folder, filename);
                    if (fs.existsSync(yamlPath)) {
                        await spawnContainer(yamlPath, ownImage);
                    }
                }
            } catch (error) {
                log('ERROR', `Error checking container ${containerName}:`, error.message);
            }
        }
    }, 30000);
    
    // Keep process alive
    process.stdin.resume();
}

// Parse legacy config file
async function parseLegacyFile(configFile, ownImage) {
    log('INFO', 'Parsing legacy configuration file...');
    
    const content = fs.readFileSync(configFile, 'utf8');
    const lines = content.split('\n');
    
    const feeds = [];
    let currentFeed = null;
    let inFeeds = false;
    
    for (const line of lines) {
        if (line.match(/^feeds:/)) {
            inFeeds = true;
            continue;
        }
        
        if (inFeeds) {
            if (line.match(/^  - name:/)) {
                if (currentFeed) {
                    feeds.push(currentFeed);
                }
                currentFeed = { name: line.replace(/^  - name:\s*/, '').trim() };
            } else if (line.match(/^    \w+:/)) {
                const [key, ...valueParts] = line.trim().split(':');
                const value = valueParts.join(':').trim().replace(/^["']|["']$/g, '');
                currentFeed[key] = value;
            }
        }
    }
    
    if (currentFeed) {
        feeds.push(currentFeed);
    }
    
    if (feeds.length === 0) {
        log('ERROR', 'No feeds defined in configuration file');
        process.exit(1);
    }
    
    log('INFO', `Found ${feeds.length} feed(s) to process`);
    
    // Spawn containers for each feed
    for (const feed of feeds) {
        await spawnLegacyFeed(feed, ownImage);
    }
    
    log('INFO', `Successfully started ${feeds.length} LDES consumer(s)`);
}

// Spawn container from legacy feed config
async function spawnLegacyFeed(feed, ownImage) {
    const feedName = feed.name || 'unnamed';
    const url = feed.url;
    const sparqlEndpoint = feed.sparql_endpoint;
    
    if (!url || !sparqlEndpoint) {
        log('ERROR', `Feed '${feedName}' is missing required 'url' or 'sparql_endpoint'`);
        return false;
    }
    
    const containerName = `ldes-consumer-${feedName}`;
    const pollingInterval = parseInt(feed.polling_interval || '60');
    const pollingFrequency = pollingInterval * 1000;
    
    log('INFO', `Starting LDES consumer for feed: ${feedName}`);
    log('INFO', `  URL: ${url}`);
    log('INFO', `  SPARQL Endpoint: ${sparqlEndpoint}`);
    
    const dockerArgs = [
        'run', '-d',
        '--name', containerName,
        '--network', DOCKER_NETWORK,
        '--label', `com.docker.compose.project=${COMPOSE_PROJECT_NAME}`,
        '--label', 'com.docker.compose.service=ldes-consumer',
        '-v', `/data/ldes-state-${feedName}:/state`,
        '-e', `LDES=${url}`,
        '-e', `SPARQL_ENDPOINT=${sparqlEndpoint}`,
        '-e', `SHAPE=${feed.shape || ''}`,
        '-e', `TARGET_GRAPH=${feed.target_graph || ''}`,
        '-e', `FOLLOW=${feed.follow || 'true'}`,
        '-e', `MATERIALIZE=${feed.materialize || 'false'}`,
        '-e', `ORDER=${feed.order || 'none'}`,
        '-e', `LAST_VERSION_ONLY=${feed.last_version_only || 'false'}`,
        '-e', `FAILURE_IS_FATAL=${feed.failure_is_fatal || 'false'}`,
        '-e', `POLLING_FREQUENCY=${pollingFrequency}`,
        '-e', `CONCURRENT_FETCHES=${feed.concurrent_fetches || '10'}`,
        '-e', `FOR_VIRTUOSO=${feed.for_virtuoso || 'false'}`,
        '-e', `QUERY_TIMEOUT=${feed.query_timeout || '1800'}`,
        ownImage
    ];
    
    try {
        await execAsync(`docker ${dockerArgs.join(' ')}`);
        
        await new Promise(resolve => setTimeout(resolve, 2000));
        
        const { stdout } = await execAsync(`docker inspect -f '{{.State.Running}}' ${containerName}`);
        if (stdout.trim() === 'true') {
            log('INFO', `Successfully started container: ${containerName}`);
            activeContainers.set(feedName, containerName);
            return true;
        } else {
            log('ERROR', `Container ${containerName} failed to start`);
            return false;
        }
    } catch (error) {
        log('ERROR', `Failed to spawn container for feed '${feedName}':`, error.message);
        return false;
    }
}

// Cleanup on exit
async function cleanup() {
    log('INFO', 'Shutting down all LDES consumers...');
    
    for (const filename of activeContainers.keys()) {
        await stopContainer(filename);
    }
    
    releaseLock();
    process.exit(0);
}

// Main function
async function main() {
    try {
        // Acquire lock
        await acquireLock();
        
        // Get own container info
        const ownInfo = await getOwnContainerInfo();
        log('INFO', `Running in container: ${ownInfo.name}`);
        log('INFO', `Using image: ${ownInfo.image}`);
        
        // Determine mode based on LDES_CONFIG_PATH
        if (fs.statSync(LDES_CONFIG_PATH).isDirectory()) {
            // Folder mode
            log('INFO', `Folder mode: watching ${LDES_CONFIG_PATH}`);
            await watchFolder(LDES_CONFIG_PATH, ownInfo.image);
        } else if (fs.statSync(LDES_CONFIG_PATH).isFile()) {
            // Legacy file mode
            log('INFO', `Legacy mode: reading ${LDES_CONFIG_PATH}`);
            await parseLegacyFile(LDES_CONFIG_PATH, ownInfo.image);
            
            // Monitor containers
            log('INFO', 'Monitoring containers... (Press Ctrl+C to stop)');
            setInterval(async () => {
                for (const [feedName, containerName] of activeContainers.entries()) {
                    try {
                        const { stdout } = await execAsync(`docker inspect -f '{{.State.Running}}' ${containerName}`);
                        if (stdout.trim() !== 'true') {
                            log('WARNING', `Container ${containerName} is not running`);
                        }
                    } catch (error) {
                        log('ERROR', `Error checking container ${containerName}:`, error.message);
                    }
                }
            }, 30000);
            
            // Keep process alive
            process.stdin.resume();
        } else {
            log('ERROR', `LDES_CONFIG_PATH is neither a file nor directory: ${LDES_CONFIG_PATH}`);
            process.exit(1);
        }
        
    } catch (error) {
        log('ERROR', 'Fatal error:', error.message);
        releaseLock();
        process.exit(1);
    }
}

// Handle signals
process.on('SIGTERM', cleanup);
process.on('SIGINT', cleanup);

// Run main
main().catch(error => {
    log('ERROR', 'Unhandled error:', error);
    releaseLock();
    process.exit(1);
});
